### ISSUE TITLE
Allow inserting PDFs as figures

### ISSUE THREAD

**Issue by MATSMACKE** [typst/typst#145]
This would allow for workflows such as creating images in `TikZ` standalone mode and then inserting them into Typst documents. This workflow would be a good stopgap until diagramming tools are integrated into Typst itself.

**Comment by daeh** [comment]
This would be critical for me to replace latex with typst. All of my paper figures are vector PDF

**Comment by user202729** [comment]
At the moment a workaround is to manually `pdf2svg a.pdf a.svg`.

**Comment by taylorh140** [comment]
It's notable that with the workaround, the text inside of svg images doesn't seem to be selectable. It makes it very noticeable when an item is included in this way, as opposed to reproducing it in raw typst.

**Comment by wrenger** [comment]
I've looked deeper into embedding pdfs into one another. Generally, pdf pages are encoded Streams that could be directly copied into XObjects. So there are two difficulties:
1. Parse the page, it's media_box (width and height...), and font references
2. Parse and copy over the embedded fonts

I played a little bit with `lopdf` and `pdf_writer` and was able to extract pages and fonts with the former and copy them into a new pdf with the latter.

In the context of typst, it might be better to either extend `pdf_writer` for the rudimentary parsing necessary or write a new small pdf parser that uses the same data structures instead of using `lopdf`.
From what it looks like, the parser would only have to be able to find pages and fonts would not need any advanced features.

**Comment by laurmaedje** [comment]
The problem is not just embedding PDFs into another. We also render directly to raster images for the web app and PNG export. As a result, supporting PDF images would mean that Typst needs to be able to render PDF files. I'm not aware of any Rust library that can do it and we can't use C libraries due to WebAssembly linking limitations. (And even if it would work, C libraries would complicate the nice, simple build process.)

**Comment by user202729** [comment]
What about PDF.js? That library can render PDF directly in the browser right? (I don't use the webapp though so not sure what's the use case here)

**Comment by MATSMACKE** [comment]
I'm not sure how Typst internals work but could it potentially be possible to reverse the code that generates PDF files to turn the PDF into some Typst intermediate representation and then rasterize through the normal method?

**Comment by wrenger** [comment]
> The problem is not just embedding PDFs into another. We also render directly to raster images for the web app and PNG export. As a result, supporting PDF images would mean that Typst needs to be able to render PDF files. I'm not aware of any Rust library that can do it and we can't use C libraries due to WebAssembly linking limitations. (And even if it would work, C libraries would complicate the nice, simple build process.)

There is, for example, https://github.com/pdf-rs/pdf_render, but I don't know how far they are.

So there are three options:
1. Just do not support pdf
2. Only support pdf embedding for pdf outputs (does not require rendering)
3. Or render/rasterize pdfs

I fear that pdf rendering might be too complex, and 3) might be out of scope for now. 2) might be a temporal solution, but would also require a decent amount of work. And for 1), the SVG conversion workaround might be sufficient for now, but it relies on third-party tools and the SVG support does definitely need some work regarding text and embedded fonts.

**Comment by laurmaedje** [comment]
> What about PDF.js?

That wouldn't work with the local CLI that can also export raster images. More generally, the image handling code shouldn't diverge between CLI and web app.

> I'm not sure how Typst internals work but could it potentially be possible to reverse the code that generates PDF files to turn the PDF into some Typst intermediate representation and then rasterize through the normal method?

It would be possible, but still very complex to implement. Note that for SVG we don't do that, we directly render it. (We have considered it and might still do it, but that would require extensions to Typst's intermediate representation.)

> I fear that pdf rendering might be too complex, and 3) might be out of scope for now. 2) might be a good enough solution, but would also require a decent amount of work. And for 1), the SVG conversion workaround might be sufficient for now, but it relies on third-party tools and the SVG support does definitely need some work regarding text and embedded fonts.

The SVG work has to be done anyway. As for the web app, we want to support PDF to SVG conversion on upload, so that it would be very simple for the user. Long-term, we _might_ support PDF, but then I would like to do it the right way. Option 2 seems like too much work for a bad user experience in the end (no preview of PDF images in the web app).

**Comment by wrenger** [comment]
> The SVG work has to be done anyway. As for the web app, we want to support PDF to SVG conversion on upload, so that it would be very simple for the user. Long-term, we _might_ support PDF, but then I would like to do it the right way. Option 2 seems like too much work for a bad user experience in the end (no preview of PDF images in the web app).

That is reasonable. I just was curious about PDF embedding from a technical standpoint, and, as PDF is **the** format for anything document related, it seems quite relevant (every tool has a at least a PDF export).
(However, I could live without it if SVG is good enough)

**Comment by tractatus** [comment]
I cant see typst being adopted for any serious academic work unless this is resolved.
All journals require figures to be submitted in PDF format. This practise is not going away anytime soon.

**Comment by LaurenzV** [comment]
As mentioned above, it's unfortunately not that easy to implement from a technical point of view.

**Comment by Andrew15-5** [comment]
I don't know if this exists, but pdf2svg package with WASM can be a good alternative for now. But without text selection.

**Comment by tractatus** [comment]
That is very unfortunate. While I really do love my short experience with typst, especially when used from Quarto where I could render plots from R and Python in one single document, and as SVG with sharp vector graphics, being able to render figures is much more important. 

For multipanel figures with raster and vector graphics combined one usually uses third-party vector graphic software to assemble the figure. Here when I export it to SVG or use pdf2svg the fact that SVG can't embed fonts is enough to create a lot of issues even in standard cases such as Arial, Helvetica, etc. on the same machine.

**Comment by laurmaedje** [comment]
Technically, SVG can embed glyph definitions. I'm not sure whether there are many tools that generate such SVGs though. Many have a "flatten" option, which converts text to paths, keeping the appearance but making the text not selectable. Currently, not even "real" SVG text is selectable in Typst, but that's something we can fix. Unfortunatley, supporting PDF images in all of Typst's output formats (PDF, PNG, SVG, HTML in the future) is _really_ complicated.

**Comment by Andrew15-5** [comment]
Related but not really useful. Today I decided to create a quick meme from a template image. Krita is (well) known for bad text editor tool, so I've seen some people say that they import text from another tool. When I did this with Inkscape, I found out that don't isn't saved in the SVG. But converting all text objects to paths fixed the issue.

Even though I don't know how tools like pdf2svg work, text to path conversion can be definitely implemented in WASM.

**Comment by Andrew15-5** [comment]
> Currently, not even "real" SVG text is selectable in Typst, but that's something we can fix.

How?

**Comment by Andrew15-5** [comment]
I also wonder how packages like `pdfpages` for LaTeX makes text selectable when inserting PDF files. 🤔

**Comment by laurmaedje** [comment]
> > Currently, not even "real" SVG text is selectable in Typst, but that's something we can fix.

> How?

The SVG tree produced by `usvg` still contains the text. We currently convert it to paths using usvg's text layout. Instead, we could use Typst's text layout to layout it (tricky with SVG features like text-along-path though, so maybe only a subset of text would be supported by this). Or we could try to somehow add text support to svg2pdf in a different way, which would likely require upstream work in usvg.

> I also wonder how packages like pdfpages for LaTeX make text selectable when inserting PDF files. 🤔

LaTeX has it far easier because it can just embed the PDF into the other PDF more or less verbatim (this is still not trivial, but doable). Then, the text stays selectable. The trouble is supporting PDF figures for all non-PDF targets (including raster image export which is used for the web app preview).

**Comment by Andrew15-5** [comment]
> Instead, we could use Typst's text layout to layout it

I briefly saw the SVG file content, but I didn't see the saved font face in it. Maybe I've missed it? How can any converter know which font to include? Do you just choose which font you want in your SVG image (when typesetting)?

> The trouble is supporting PDF figures for all non-PDF targets

I only really use PDF, so I don't know why you can't just `pdf2svg` or `pdfimages -png` it? Is it because it's much slower than Typst compilation?

I also don't know if this is a good move, but can't we add support for PDF embedding only for PDF target (for now)? Yes, the online app won't support this (I personally prefer using Typst locally). But most documents are used as PDF anyway (and all those journal things).
1. Typst is in beta anyway, so funky stuff, bugs and breaking changes are OK (in general)
2. Adding limited support (only from PDF) of this feature will greatly increase positive feedback from all those users who want to embed PDF in their own PDF documents.

**Comment by Andrew15-5** [comment]
Special Unicode characters like colored emojis are still not available in PDF (#144), but are in other targets, right? So why we can't _compensate_ this by giving PDF a feature that isn't (ATM) available for other targets?

**Comment by Enivex** [comment]
> I cant see typst being adopted for any serious academic work unless this is resolved.
> All journals require figures to be submitted in PDF format. This practise is not going away anytime soon.

I feel like this is putting the cart before the horse. Unless it's a journal that accepts PDF submissions without a LaTeX source (which I can't imagine there are many of), you're not submitting a typst document anytime soon, regardless. Moreover, you can always attach the figures in PDF format as well, if the journal only accepts that for whatever reason.

**Comment by tractatus** [comment]
@Enivex I do not agree. As an academic who published, reviewed, and served in editorial roles I can say that converting to PDF and even PDF to PDF is an intricate part of how the publishing system is built. We can think all we want about whether this is correct or not. But just like the financial industry is built on top of some legacy COBOLT code so is the academic publishing industry built on top of PDF.

Most journals that allow TeX submissions will tell you fairly quickly into the revision process to just send them PDF instead.

> Unless it's a journal that accepts PDF submissions without a LaTeX source (which I can't imagine there are many of)

Pretty much all journals accept PDF submissions without a TeX source. I never heard of any journal that will require you to submit a TeX file to supplement your PDF, they do not care about your TeX file, it creates more hassle than simply using the PDF tools they already have.

**Comment by ziprandom** [comment]
For anyone that looks for a workaround, this script wraps `typst` and identifies file paths that match either: `<some_filename>.pdf.svg` or `<some_filename>.pdf.<pagenumber>.svg`, and converts the respective `<some_filename>.pdf` to the found filename extracting page `<pagenumber>` or 1. 

This works nicely with the `compile` subcommand.

problems: 
* will only include one (default first) page of the pdf. so to include more pages, you will need something like: 
```
#image("pdfs/my_pdf.1.svg")
#image("pdfs/my_pdf.2.svg")
#image("pdfs/my_pdf.3.svg")
```
* will parse the source file so programmatically created links will not work
* the `watch` command will not pick up changes to the included pdfs. Also only one page of the pdf will be included at a time, so every page will have to be included manually: 

* `typst.sh` (`chmod +x`)
```sh
#!/usr/bin/env bash
TARGET="${@: -1}"
PDFS_TO_CONVERT=$(cat $TARGET | grep -oP '[^"]+\.pdf(\.[0-9]+)?\.svg?')
declare -a OUTPUT_FILES
for INPUT in $PDFS_TO_CONVERT; do
  SEGMENTS=($(sed -r 's/([^"]+.pdf)\.([0-9]+)?.?svg/\1 \2/g' <<< ${INPUT}))
  PDF=${SEGMENTS[0]}
  IDX=${SEGMENTS[1]}
  OUT_FILE_NAME=""
  if [[ ! -z "${IDX}" ]];
  then
    OUT_FILE_NAME=${PDF}.${IDX}.svg
  else
    OUT_FILE_NAME=${PDF}.svg
  fi
  pdf2svg ${PDF} ${OUT_FILE_NAME} ${IDX}
  OUTPUT_FILES+=(${OUT_FILE_NAME})
done;
typst "$@"
rm ${OUTPUT_FILES[@]}
```

**Comment by user202729** [comment]
Worth noting that now `typst query` exists, so maybe you can consider using that instead of `sed` etc.

**Comment by ziprandom** [comment]
I looked into that but didn't see it beeing able to query anything other
than metadata. Would be really nice to use it though.

**Comment by laurmaedje** [comment]
It can query other stuff than metadata (it can query the same things the `query` function can). However, the list of things is still limited because making everything queryable needs some compiler work to be efficiently possible. Note that any element becomes queryable if you attach a label to it.

**Comment by user202729** [comment]
> I looked into that but didn't see it beeing able to query anything other than metadata. Would be really nice to use it though.

Metadata is, **by definition**, anything that can be queried. So anything you want to be queried, you pass it to the metadata.

So the plan is in the `image()` function, you print to the page a (invisible) [content] block that is produced by `#metadata("/path/to/pdf.pdf") #label("pdftosvglabel-"<some counter here>)`, then parse the output of `typst query` for the labels that start with `pdftosvglabel` and do the work needed.

Perhaps the Typst documentation could have some tutorial/example to illustrate how feature can be used, instead of just the formal definition.

**Comment by laurmaedje** [comment]
I agree that it's hard to grasp at the moment. I want to [rework](https://laurmaedje.github.io/posts/types-and-context/) it and then also have better docs for it.

**Comment by billchen2k** [comment]
+1 for this. I export all my figures in PDF in my previous LaTeX workflow and I'd really like to see typst supporting this.

**Comment by raievsky** [comment]
Asciidoctor-pdf does this. It might be interesting to look at how they did that.

**Comment by janosh** [comment]
Is there a way to crowdsource a bounty for this issue?

**Comment by Andrew15-5** [comment]
> Is there a way to crowdsource a bounty for this issue?

Maybe there is/will be a dedicated place for this, but IMO you can just donate/sponsor Typst on GitHub one time here: https://github.com/sponsors/typst?frequency=one-time. But the official members must state if they're willing to change priority of some tasks/issues if a bounty is placed on a specific one. I'm not one of them.

Right now, this is the only place where you can sponsor Typst.

Note that a lot of work must be done to ["styling rework"](https://discord.com/channels/1054443721975922748/1175894504146993304) in order to fix a lot of bugs connected to it. So it may so happen that right now, setting a bounty wouldn't make much of a difference.

**Comment by user202729** [comment]
Well, Typst (core) is open source so if you can obviously hire a programmer to implement the feature (if you can find one willing). (someone might also implement a hack to do `pdf2svg` internally though, which is not too hard to implement given "including" Python is also possible)

**Comment by Andrew15-5** [comment]
> (someone might also implement a hack to do `pdf2svg` internally though, which is not too hard to implement given "including" Python is also possible)

Inside Typst you can run Python 3.12 code via pyrunner package (which runs inside WASM), but you can't load 3rd party Python libraries, unless you read necessary files directly through Typst. Running Python inside Rust is more complex, I think.

**Comment by janosh** [comment]
One nice thing about PDF figures is they're typically lower file size than SVG. Using `pdf2svg` defeats that purpose. As a v0 solution, it would still be nice to have.

**Comment by LaurenzV** [comment]
The reason that this hasn't been implemented yet is that this would require us to write us a PDF renderer in pure Rust. Doing this properly is probably _at least_ a something that will take many months, if not even years, assuming that you have a few developers working on it full-time. 

PDF unfortunately is such a complex format that implementing this is anything but trivial.

There is some hope that at some point we might be able to create a WASM plugin that calls mupdf (which is a battle-tested PFF processor that works really well) in the background to convert it to SVG, but even if that happens, it would mean that text is converted to paths, which isn't that great either... But it would be a start.

**Comment by LaurenzV** [comment]
Yes, it's possible to convert to an SVG using mupdf, and it works pretty well from what I've tried.

mupdf also has an option to preserve text, but at least from my experiments the text looks really janky, so I don't think it's a good option.

**Comment by Enivex** [comment]
The problem is that mutool / mupdf currently can't become plugins due to lack of exceptions in wasm

**Comment by reknih** [comment]
Note that the web application will automatically convert PDFs to SVGs when uploaded so that they can be used with Typst. This feature has been added last month.

**Comment by janosh** [comment]
@reknih That's a great feature! I just tried it out and works great. Out of curiosity, I downloaded the converted SVG to check its size. It's 264 KB, more than 10x the original 24 KB PDF figure. I then uploaded the Typst-converted SVG to https://vecta.io/nano which managed to compress the SVG to 96 KB.

![Screenshot 2024-01-01 at 6 29 09 PM](https://github.com/typst/typst/assets/30958850/0dc3fd21-52ff-4172-99bd-81f73966594c)


All this to say how about compressing auto-converted SVGs with something like `svgo` on the fly?

**Comment by mhdadk** [comment]
> > The problem is not just embedding PDFs into another. We also render directly to raster images for the web app and PNG export. As a result, supporting PDF images would mean that Typst needs to be able to render PDF files. I'm not aware of any Rust library that can do it and we can't use C libraries due to WebAssembly linking limitations. (And even if it would work, C libraries would complicate the nice, simple build process.)
> 
> There is, for example, https://github.com/pdf-rs/pdf_render, but I don't know how far they are.
> 
> So there are three options:
> 
>     1. Just do not support pdf
> 
>     2. Only support pdf embedding for pdf outputs (does not require rendering)
> 
>     3. Or render/rasterize pdfs
> 
> 
> I fear that pdf rendering might be too complex, and 3) might be out of scope for now. 2) might be a good enough solution, but would also require a decent amount of work. And for 1), the SVG conversion workaround might be sufficient for now, but it relies on third-party tools and the SVG support does definitely need some work regarding text and embedded fonts.

Just wondering, for option 3, would [servo/pathfinder](https://github.com/servo/pathfinder) be helpful in achieving it? I found `servo/pathfinder` under the "Renderer and Viewer" section [here](https://github.com/pdf-rs/pdf#renderer-and-viewer).

**Comment by reknih** [comment]
@janosh The file size blow-up is something that SVGO cannot fully fix because of the way we need to handle text here: PDFs embed fonts that define each character once and then print characters from the font. In SVG, we need to use each character's vector path / shape instead because we cannot make the font from the PDF available. The information (and thus compressability) that these shapes ever were text gets lost here. This is why the SVGs will be fundamentally larger. (Some of) SVGOs optimizations drop accuracy by rounding coordinates off so I'm not sure that that's something that should be on by default.

@mhdadk Pathfinder is just one of the potential components that would be needed to implement this. We feel like implementing option 3 would take a long time, occupy too much of our resources, and potentially bloat the Typst binary.

**Comment by bwood4** [comment]
> So there are three options:
> 
> 1. Just do not support pdf
> 2. Only support pdf embedding for pdf outputs (does not require rendering)
> 3. Or render/rasterize pdfs
> 

For option 2, what if for SVG/image outputs it converted PDF attachments to SVG and rendered them that way (maybe on compile rather than on upload?), then for PDF output it embeds them directly? Maybe for PDF outputs, do the SVG conversion by default and add an option to embed them directly, that way the outputs behave the same by default.

I think this would keep the web and CLI interfaces functionally equivalent, allow for direct PDF embedding (which is a must for a lot of people), and not require typst to render a PDF directly.

**Comment by laurmaedje** [comment]
The problem is that converting PDF to SVG is pretty similar in complexity to rendering a PDF. The conversion on upload happens in our backend so that Typst doesn't have to embed a PDF renderer/converter.

**Comment by bwood4** [comment]
That makes sense. It's a shame though, PDF inclusion is a pretty important feature that LaTeX has that Typst won't. How feasible would it be to create a package that allows it for PDF outputs only similar to LaTeX pdfpages?

**Comment by Andrew15-5** [comment]
> that Typst won't

![image](https://github.com/typst/typst/assets/37143421/84b87e85-15fe-4f48-8d6a-6d934f81d485)

**Comment by Enivex** [comment]
> https://www.ams.org/publications/journals/journalsframework/jamssubmit
> 
> _"The Journal of the AMS has a single anonymous process in which the reviewers know who the authors of the manuscript are, but the authors do not have access to the information on who the peer reviewers are. The AMS uses Centralized Manuscript Processing for initial submissions to AMS journals. **The preferred method of submission is to upload a PDF file** using the [Initial Manuscript Submission form](https://ams.msp.org/submit_new.php?j=jams)."_

Initial submission, yes, but not after the paper has been accepted for publication.

**Comment by daskol** [comment]
As a workaround solution until Typst implements support for PDF/SVG figures, it is possible to directly render figure to Typst drawing primitives like `path`, `move`, `text, etc. This is similar to PGF graphics in TeX. Figure can be rendered in a separate `*.typ` file and included in the main body later.

```typst
#figure(
  include "path/to/my/fancy/figure.typ",
  kind: image,
  caption: [Caption for my fancy figure.],
) <line-plot-simple>
```

Specifically, I implemented [custom rendering backend][1] for Python's `matplotlib`. It might be useful for those who prepares figures in Python. My custom backend does not implement rendering raster objects but the rest work well. I prepared about 10 figures for my paper. It's initial compile time is about 3 seconds but subsequent compilations take casual 50 ms. Another drawback is that the custom backend invokes `typst` binary directly with caching that increases rendering time for the first run.

[1]: https://github.com/daskol/typst-mpl-backend

**Comment by Enivex** [comment]
> As a workaround solution until Typst implements support for PDF/SVG figures

Typst has supported SVG since forever

**Comment by daskol** [comment]
> Currently, not even "real" SVG text is selectable in Typst, but that's something we can fix.

> As a workaround solution until Typst implements support for PDF/SVG figures

Agreed. To some extent.

**Comment by daeh** [comment]
Just wanted to again suggest that support for PDF figures should be a priority. Huge thanks to @laurmaedje and the other devs for all their work.

**Comment by wrenger** [comment]
I just did a bit of prototyping on how to embed PDFs (using lopdf+pdf-writer): https://github.com/wrenger/pdf-embedding

Even though this prototype is not perfect, the general procedure (which is inspired by latex) is not that complicated.
Just parse the desired page (Content+Resources) and copy it into a new `FormXObject`. Then copy all indirect objects/streams that have been referenced by the Resources (also renumber the references).

For parsing, the [lopdf](https://github.com/J-F-Liu/lopdf) library does most of the heavy lifting, but it has two downsides. It parses the whole PDF (not only the desired page), which is a problem in my opinion, and it does many things we don't need. Maybe we are better off writing our own small parser (using parts of lopdf like [parser.rs](https://github.com/J-F-Liu/lopdf/blob/master/src/parser.rs)), and then only read and parse the objects we need.

**Comment by Enivex** [comment]
> I just did a bit of prototyping on how to embed PDFs (using lopdf+pdf-writer): https://github.com/wrenger/pdf-embedding
> 
> Even though this prototype is not perfect, the general procedure (which is inspired by latex) is not that complicated.
> Just parse the desired page (Content+Resources) and copy it into a new `FormXObject`. Then copy all indirect objects/streams that have been referenced by the Resources (also renumber the references).
> 
> For parsing, the [lopdf](https://github.com/J-F-Liu/lopdf) library does most of the heavy lifting, but it has two downsides. It parses the whole PDF (not only the desired page), which is a problem in my opinion, and it does many things we don't need. Maybe we are better off writing our own small parser (using parts of lopdf like [parser.rs](https://github.com/J-F-Liu/lopdf/blob/master/src/parser.rs)), and then only read and parse the objects we need. 

As has been stated many times, the challenge is svg and png output.

**Comment by wrenger** [comment]
> As has been stated many times, the challenge is svg and png output.

I know, rendering PDFs to svg/png is currently out of scope for Typst (and this will probably be the case for a long time). 

Still, many want to embed PDFs into PDFs, especially those coming from LaTeX and Tikz, who have a lot of PDF figures. I've had countless discussions with peers that want this and think Typst is *not* mature enough because it doesn't support PDF embedding.

Also, I think that for PDF into PDF embedding, this direct copy-paste solution would actually be still viable, even if we were able to render a PDF. It is just much faster and accurate. Only for SVG/PNG outputs, I would render the PDF.

**Comment by Enivex** [comment]
> Still, many want to embed PDFs into PDFs, especially those coming from LaTeX and Tikz, who have a lot of PDF figures. I've had countless discussions with peers that want this and think Typst is not mature enough because it doesn't support PDF embedding.

Tools like mutool can convert pdf to svg. For _most_ figures that'll be more or less lossless. This functionality is already integrated into the web app.

Anyway, the devs are hesitant about including functionality that is specific to a single output format. This one would be especially problematic because it wouldn't show up in the preview of either the web-app or typst-preview (vscode), since those aren't pdf based.

**Comment by LaurenzV** [comment]
> Also, I think that for PDF into PDF embedding, this direct copy-paste solution would actually still be viable, even if we were able to render a PDF. It is just much faster and accurate. Only for SVG/PNG outputs, I would render the PDF.

I think no one is questioning that. Even if we already had PDF rendering, the rendering would only be used for SVG/PNG output, while for PDF we would just, as you mentioned, "copy paste" the relevant objects into a new PDF file. The way this would work is more or less like you described (although it is a bit more trickier, because you not only need to find the references objects in the main page stream, but also in all recursively in all other references objects).  But this is definitely doable and I would estimate it's probably a couple of week's worth of work if you know what you're doing. If there is interest in that, I could attempt working on this during my summer vacation.

But as was mentioned above, at least currently it does not seem to be planned to add this until we have a decent solution for the other output formats, so I'm not sure it's worth spending time on this for now. Even though it seems to be one of the most-requested features.

**Comment by wrenger** [comment]
> although it is a bit more trickier, because you not only need to find the references objects in the main page stream, but also in all recursively in all other references objects

Yes, the [prototype](https://github.com/wrenger/pdf-embedding) already does this, take a look :D
I plan to integrate this into Typst and make a PR in a few days.

Regarding the other output formats, you already use external tools for that in the web app. How about adding a flag to the Typst cli to specify an external tool (pdf2svg, mutool...) to convert PDFs to SVGs.
With this, `#image("foo.pdf")` would be either directly embedded or converted into an SVG, depending on the output.

**Comment by wdscxsj** [comment]
My deep feeling is that many users expect a better (easier, faster, more flexible, etc.) LaTeX in the first place, and PDF output is the gold standard. Thus including PDF figures as-is is a killer feature, despite the design decision that the DTP end (PDF output) and the Web end (SVG and PNG output) should have the same effect.

Personally this issue is the only thing that prevents me from switching to Typst for serious work. SVG and PDF are both complex, indeed, but PDF is self-contained (especially for fonts), well supported everywhere, and the first (if not the only) thing that many LaTeX users care about. It deserves some special love, as indicated by the high votes.

**Comment by laurmaedje** [comment]
> Regarding the other output formats, you already use external tools for that in the web app. How about adding a flag to the Typst cli to specify an external tool (pdf2svg, mutool...) to convert PDFs to SVGs.

The Typst compiler should be portable, safe, and easy to use. Shelling out to an external program is problematic for all three of these things. With the web app, it's quite different, since the PDF to SVG functionality has really nothing to do with Typst itself, it is a one time backend task that happens on upload.

> I plan to integrate this into Typst and make a PR in a few days.

I admire the effort, but ultimately, as discussed, the PDF to PDF embedding isn't the problem. As such, a PR like that wouldn't really change anything about the status quo.

**Comment by wrenger** [comment]
> I admire the effort, but ultimately, as discussed, the PDF to PDF embedding isn't the problem. As such, a PR like that wouldn't really change anything about the status quo.

Well then, what is the plan here? If we agree that this is the right way for PDF to PDF embedding, how long do we want to wait for this to be implemented?

I'm just looking for a compromise. Falling back to an external program is not ideal, but for an edge-case (PDF to SVG/PNG), this might be the best option for now. Maybe in the future, Typst can have an integrated converter...

**Comment by Enivex** [comment]
> PDF output is the gold standard

PDF figures being the gold standard is largely because it's the only vector format pdflatex can handle.

> Web end (SVG and PNG output)

It's not just web. Decent real time preview doesn't rely on pdf either.
 
> Personally this issue is the only thing that prevents me from switching to Typst for serious work.

Speaking as a long time LaTeX user, converting PDF to SVG with mutool usually works very well. 

> but PDF is self-contained (especially for fonts), well supported everywhere

SVG can also be self-contained, and is supported perhaps even more places than PDF is.

(Yes I'm aware that some tools unfortunately produce SVGs with foreignObjects in them, making those useless outside of web browsers)

> and the first (if not the only) thing that many LaTeX users care about.

Are they aware that SVGs can serve the same purpose?

> It deserves some special love, as indicated by the high votes.

It's not a question of will. There simply just doesn't exist a viable path for PDF figures in typst for the foreseeable future.

The most promising thing on the horizon would probably be compiling mutool to WASM allowing pdf -> svg on the fly as a plugin. However, even that requires WASM 2.0, which isn't yet finalized.

**Comment by Enivex** [comment]
> > I admire the effort, but ultimately, as discussed, the PDF to PDF embedding isn't the problem. As such, a PR like that wouldn't really change anything about the status quo.
> 
> Well then, what is the plan here? If we agree that this is the right way for PDF to PDF embedding, how long do we want to wait for this to be implemented?
> 
> I'm just looking for a compromise. Falling back to an external program is not ideal, but for an edge-case (PDF to SVG/PNG), this might be the best option for now. Maybe in the future, Typst can have an integrated converter...
> 

The plan would be to rely on SVG figures until there is a viable solution. I don't think a compromise is going to happen.

**Comment by wdscxsj** [comment]
@Enivex Sorry, I'm not quite sure here, but are you speaking for the Typst team?

**Comment by Enivex** [comment]
> @Enivex Sorry, I'm not quite sure here, but are you speaking for the Typst team?

No I am not

**Comment by AndreKR** [comment]
> Speaking as a long time LaTeX user, converting PDF to SVG with mutool usually works very well.

Wait, I only just now understood that simply converting the PDF to SVG using `mutool convert` is a totally feasible workaround.

It was pretty hard to find `mutool` though. For Windows at least you have to go to https://mupdf.com/releases, scroll down to version 1.23.0, download `mupdf-1.23.0-windows.zip` and in there you will find `mutool`.

**Comment by Andrew15-5** [comment]
I only ever used [`pdf2svg`](https://cityinthesky.co.uk/opensource/pdf2svg/) since it's the first and only tool I found. Need to compare it to `mutool`, but I don't think there are any reasons to switch (for me).

**Comment by cgahr** [comment]
Personally, I don't mind using svg over pdf, but for one issue: In pdf figures, embedded into latex, I can search the figure labels with Ctrl+F. As far as I can tell, this is not the case for typst and I cannot search the text in svg figures. 

That's really the only issue I have.

**Comment by LaurenzV** [comment]
This will actually be possible with the next update!

**Comment by YDS67** [comment]
If svg figures can be self-contained and would allow for text search and copy, then I'll have no issue with using svg instead of pdf. Asymptote creates svg files which are identical to pdf files, so it's all good to me. Here's an example figure with equations generated by Asymptote. If Typst ever gets a figure generating tool as powerful as Asymptote, I might switch to it completely.

![first](https://github.com/typst/typst/assets/76599593/49155fc8-31ed-4f20-905c-672caaf42009)

**Comment by user202729** [comment]
@YDS67 Unrelated but you can just use Asymptote. https://github.com/patrick-kidger/typst_pyimage already exists and I imagine it wouldn't be hard to adapt it to Asymptote.

Side note but Asymptote uses TeX to render formulas still. Would be nice if it has two-way integration (use Typst to render formula).

**Comment by tobiasdiez** [comment]
> > I plan to integrate this into Typst and make a PR in a few days.
> 
> I admire the effort, but ultimately, as discussed, the PDF to PDF embedding isn't the problem. As such, a PR like that wouldn't really change anything about the status quo.

It would solve the problem for all users that ultimately only care about the final PDF output. I agree, it would be awesome to properly render the embedded PDF in the webapp and/or vscode preview, but at least for me it would be totally fine to display a simple placeholder in these cases. I already know what is in the embedded pdf, and most often only its alignment relative to other elements of the page is important.

Personally, I also don't quite understand the problem. It seems like a special handling of embedded PDFs for PDF output is necessary either way. So why not start implementing this comparably simple case, and then later provide proper support for other output formats?

**Comment by Enivex** [comment]
> Personally, I also don't quite understand the problem. It seems like a special handling of embedded PDFs for PDF output is necessary either way. So why not start implementing this comparably simple case, and then later provide proper support for other output formats?

Even assuming you're not directly interested in png or svg output, it would mean that the preview in the web app or typst-preview would not be representative of the final pdf. That's a non-starter.

**Comment by user202729** [comment]
@Enivex I don't think that reflects the view of the Typst developers

> I think no one is questioning that. Even if we already had PDF rendering, the rendering would only be used for SVG/PNG output, while **for PDF we would just, as you mentioned, "copy paste" the relevant objects into a new PDF file**.

But I don't understand what is the problem either.

**Comment by Enivex** [comment]
> @Enivex I don't think that reflects the view of the Typst developers
> 
> > I think no one is questioning that. Even if we already had PDF rendering, the rendering would only be used for SVG/PNG output, while **for PDF we would just, as you mentioned, "copy paste" the relevant objects into a new PDF file**.
> 
> But I don't understand what is the problem either.

That reply doesn't contradict what I said. _Once_ there is PDF rendering that can be used for inclusion in PNG and SVG (and thus enabling preview), then inclusion in PDF would be comparatively easy.

This doesn't change the fact that they (to my understanding) are unwilling to allow this before PDF rendering is possible.

**Comment by ruifengx** [comment]
I don't really feel this conversation is going anywhere. Clearly the two parties involved have contradicting appeals: it seems Typst takes feature parity between different backends very seriously. And it is legitimate: the software itself is open-source, so only the online service is practically profitable; and ensuring profits is the most effective way to guarantee long-term maintenance. I would not expect the developers to actively work on a feature that makes the online editing experience inferior (recall that WASM extensions won over native extensions, and one of the reasons was portability, despite the performance/size issues for e.g. [the tree-sitter-based highlighting plugin](https://github.com/typst/packages/pull/143)), and I find their concerns completely legitimate and understandable.

**Comment by reknih** [comment]
> pdf2svg as wasm when?

People can package it as a plugin. The software is GPL-v2 licensed which may or may not create problems. At any rate, we cannot integrate it.

**Comment by drskoolie** [comment]
I keep on coming back to this. 

I've written a python script that:
1. takes in PDFs and converts them to SVGs
2. then writes the typst text to import the SVGs as images with the correct size for the page. 


Problem is, the PDF -> SVG takes a long time (Using PyMuPDF) and it balloons the PDF. **A 3MB PDF becomes 50MB.** 

Has anyone found a way to solve this? I cannot use PNGs because I need my final PDF to be text searchable.

**Comment by AndreKR** [comment]
@drskoolie The name "PyMuPDF" suggests that it might be a binding for the MuPDF library, but to rule out any contribution of Python to the long runtime could you try converting it with `mutool`?

I also noted that some SVGs are much larger than the equivalent PDFs. In PDFs the contained streams are compressed, while I think there is no such mechanism in SVG. It seems common to gzip SVGs when they are served by a web server. Maybe Typst can also read gzipped SVGs?

**Comment by Kreijstal** [comment]
You would have to compile mutool to wasm so it runs on the browser

**Comment by janosh** [comment]
be sure to run `svgo --multipass path/to/pdf-converted-to.svg`. the resulting SVG is usually still larger than the original PDF but rarely by as much as 3MB->50MB. another option is vecta nano, not a CLI so more manual work involved but gives smaller SVGs than `svgo --multipass` in my experience by uploading to https://vecta.io/nano

**Comment by drskoolie** [comment]
@AndreKR 

I used `mutool` and I get the same that PyMuPDF is getting. It's not the same exact size, some are 10% bigger and some are 10% smaller. But overall, it's the same size.

@janosh 

I used `svgo --multipass` and it brought down the SVG size by about 50%

-----------------------------------------------------------------
Overall, it does work. But I'm not happy with the setup. SVGs take too long to convert and I'll need to batch the process so it runs in parallel. 

**There is no point for Typst to take 30ms to compile and have PDFs take 10 seconds to "compile".** 

I've thought of writing a script where I insert PDFs into typst after typst compiles. This would entail putting in blank pages in typst to maintain the correct page numbers for the table of contents, but then I lose my headers and footers. I use my footers for page numbers so it's kind of important.

**Comment by daskol** [comment]
I guess that the main cause of SVG bloating in your cases is text objects which often require implicit conversion from glyph to path objects. As a workaround one can remove text objects from PDFs/SVGs and then render texts directly with `text/move/place` in Typst.

**Comment by Enivex** [comment]
> I keep on coming back to this.
> 
> I've written a python script that:
> 
>     1. takes in PDFs and converts them to SVGs
> 
>     2. then writes the typst text to import the SVGs as images with the correct size for the page.
> 
> 
> Problem is, the PDF -> SVG takes a long time (Using PyMuPDF) and it balloons the PDF. **A 3MB PDF becomes 50MB.**
> 
> Has anyone found a way to solve this? I cannot use PNGs because I need my final PDF to be text searchable.

How complicated is this PDF exactly? My experience is that mutool is pretty much instant.

**Comment by wrenger** [comment]
> Problem is, the PDF -> SVG takes a long time (Using PyMuPDF) and it balloons the PDF. A 3MB PDF becomes 50MB.

Ideally, you would not convert pdfs to svgs on every build. Only if the pdfs have changed. E.g. exactly what `make` does...

**Comment by nickb937** [comment]
The thing I wanted was to have a PDF page as a background image and then use Typst to overlay text on-top e.g. form filling in text-boxes with word wrap.  Similar can be done with LaTeX and the tikzpicture library.  But it seems the approach I'll would have to take is to use Typst to write data to a blank page and then use a pdf-merging tool to merge the pages togther, similar to how a watermark is applied to the page.

**Comment by Enivex** [comment]
> The thing I wanted was to have a PDF page as a background image and then use Typst to overlay text on-top e.g. form filling in text-boxes with word wrap.  Similar can be done with LaTeX and the tikzpicture library.  But it seems the approach I'll would have to take is to use Typst to write data to a blank page and then use a pdf-merging tool to merge the pages togther, similar to how a watermark is applied to the page.

No, you would just have to convert the pdf to svg

**Comment by balping** [comment]
### Typst PDF embedder

I created a [python wrapper](https://gitlab.com/balping/typst-pdf-embedder) around `typst` that makes it possible to embed pdf files  **while conserving vector graphics and selectable text**. No conversion to svg is needed.

It uses the following syntax:

```typst
#figure(
  pdf("figures/plot.pdf", width: 100%),
  caption: [Full width pdf]
)
```

You can find a demo source `.typ` and the corresponding pdf file generated by my package [here](https://gitlab.com/balping/typst-pdf-embedder/-/tree/master/test).

Anyone interested in embedding pdfs, please [check it out](https://gitlab.com/balping/typst-pdf-embedder). I only tested it on Linux, but it should work on other platforms too, I welcome bug reports.


---

### How does it work?

The compile process consists of five steps:

**1.** `typst-pdf.py` generates a list of pdf files to be included later by running `typst query` that extracts metadata inserted by calls to `pdf()`.

**2.** Now `typst-pdf.py` has a list of all the pdf files. It checks the sizes of all these pdfs and generates a temporary json file, later read by `typst`, containing this size information. This is necessary, so that `typst` has knowledge about the aspect ratios of the images, so that it can reserve appropriate amount of space for them.

**3.** `typst-pdf.py` calls `typst compile`. Now typst knows how large each pdf should be and generates placeholder rectangular frames:

![sh1](https://gitlab.com/-/project/61512314/uploads/112c4880d6c5a35fa5fae535682227c8/sh1.png)

**4.** Now that the output document is correctly laid out, `typst query` is called again to obtain the actual sizes and positions of the placeholders.

**5.** The placeholder frames are identified and replaced by the embedded pdfs using PyMuPDF:


![sh2](https://gitlab.com/-/project/61512314/uploads/dd4ed7e42a65e1bf70cf83ee9e7067db/sh2.png)

**Comment by wrenger** [comment]
I was thinking about making a similar plugin with MuPDF and WASM so that it can be used in the web editor and with LSPs (like tinymist)... But there is just no time currently.

**Comment by Enivex** [comment]
> I was thinking about making a similar plugin with MuPDF and WASM so that it can be used in the web editor and with LSPs (like tinymist)... But there is just no time currently.

It's not possible to do that yet

@Dherse

**Comment by Dherse** [comment]
Don't get me wrong, it's possible, it would just require (a lot) of rust code in Typst just to support it so it wouldn't really be a plugin, more like an "extension" with all of the schafolding built-in. One we have something like WASM exceptions and unwinding, it should be fairly easy to build as a regular plugin :)

**Comment by wrenger** [comment]
Isn't mupdf a C codebase. So there shouldn't be any exceptions. Also, parts of it are already available as WASM libraries.

**Comment by Dherse** [comment]
> Isn't mupdf a C codebase. So there shouldn't be any exceptions.
That's simply not true unfortunately. The way that mupdf handles subtle differences between PDFs produced by different toolchains is that is will unwind when it hits something unexpected and then try a different code path.

But if you don't believe me, go and try it, you'll see quickly enough that mupdf does indeed use stack unwinding and is no where near being wasm-native without a boatload of boostrap JS code you simply cannot reproduce in a typst plugin at the moment.

Here is the file where they do it all: https://github.com/ArtifexSoftware/mupdf/blob/336ee9e30cfc2bd59bf75320ce2c4455aa1efd42/include/mupdf/fitz/system.h#L133

**Comment by balping** [comment]
@wrenger 

Even if you managed to compile muPDF to WASM somehow, it wouldn't help much as typst's plugin system does not give you direct access to the output file or to its document writer. From a plugin function you can merely return byte strings that should be processable by normal typst. Your only option would be to convert all pdf graphics primitives to typst source code (very tedious, not even speaking of copying fonts, which I think is currently impossible). This is essentially pdf -> typst conversion, not pdf embedding.

**Comment by wrenger** [comment]
> @wrenger 
> 
> 
> 
> Even if you managed to compile muPDF to WASM somehow, it wouldn't help much as typst's plugin system does not give you direct access to the output file or to its document writer. From a plugin function you can merely return byte strings that should be processable by normal typst. Your only option would be to convert all pdf graphics primitives to typst source code (very tedious, not even speaking of copying fonts, which I think is currently impossible). This is essentially pdf -> typst conversion, not pdf embedding.

I would produce an svg string with mupdf and put it into image.decode... 
At least this was the idea.

**Comment by balping** [comment]
> I would produce an svg string with mupdf and put it into image.decode...
> At least this was the idea.

Ok, but then we are back to all the problems of pdf to svg conversion if you read back in this thread. Problems with fonts, text non selectable, increased filesize, etc.

**Comment by Enivex** [comment]
> > I would produce an svg string with mupdf and put it into image.decode...
> > At least this was the idea.
> 
> Ok, but then we are back to all the problems of pdf to svg conversion if you read back in this thread. Problems with fonts, text non selectable, increased filesize, etc.

Actual PDF embedding won't happen in the foreseeable future due to factors that have been mentioned several times in this thread

**Comment by flosse** [comment]
> Actual PDF embedding won't happen in the foreseeable future due to factors that have been mentioned several times in this thread

I just came across this thread because I'm having trouble embedding SVG and thought it would be much easier if I embedded PDFs, but that doesn't seem to be a good idea :rofl:

Nevertheless, typst is a great project, thank you very much! I have great respect for your work!

**Comment by Enivex** [comment]
> @balping Is it possible to embed interactive content like animated GIFs with your utility into final PDF?

PDF does not support animated GIFs, so that would be impossible.

There is however the possibility of adding videos, though support for this in PDF readers is extremely poor.

**Comment by Andrew15-5** [comment]
Video support issue: https://github.com/typst/typst/issues/955.

**Comment by daskol** [comment]
Thanks.

> PDF readers is extremely poor.

Yeah. No wonder support is poor. 😢 There are better presentation tools like browser or media player. Video in PDF is indeed too much.

**Comment by jordantgh** [comment]
>There simply just doesn't exist a viable path for PDF figures in typst for the foreseeable future

>Actual PDF embedding won't happen in the foreseeable future

@Enivex seeing as you have said you don't speak on behalf of the Typst team, I don't understand why you nevertheless keep replying here to pour cold water on suggestions *as though* you were doing so. Maybe you can clarify.

Laurenz said:

>I admire the effort, but ultimately, as discussed, the PDF to PDF embedding isn't the problem. As such, a PR like that wouldn't really change anything about the status quo.

The way I read it, this implies he is of the opinion the change won't make a big difference for users. However, based on the response in this thread, and as @tobiasdiez  pointed out, it seems to me like many people would actually benefit from this. 

I can understand why the Typst team may not prioritise this change if they feel it only benefits a handful of users. However, it's worth drawing a distinction between "shouldn't implement because 'breaking the relationship between the pdf and the preview in the web-app or typst-preview is a complete non-starter'" (your claim) and "won't implement because it's not high impact enough" (something that can possibly be changed if enough people care).

**Comment by Enivex** [comment]
@jordantgh I'm merely conveying my understanding from various discussions. I should perhaps have elaborated, but this thread is not particularly productive at this point.

**Comment by reknih** [comment]
The problem is not that we think it will benefit too few users. The problem is that Typst has more targets than just PDF: PNG, SVG, and soon HTML.

To embed PDF in especially PNG or SVG, we would need to render it. For this, we need a Rust implementation of a PDF viewer which does not exist. We think that creating one is too much effort and we are not willing to add an image format that won't work in all export targets, hence @laurmaedje's comment.

**Comment by tobiasdiez** [comment]
@reknih I've no rust experience but doesn't https://github.com/ajrcarey/pdfium-render fit the bill? Even supports WASM for running in the browser.

Could you maybe also outline a strategy to implement PDF support so that people that would like to help have a starting point. For example, you mentioned that all export targets need to be supported. On the other hand, some of the problems discussed above are specific to certain export target - for example, you don't really care about embedded fonts and selectable text when exporting to PNG. So would it be okay to have an implementation that depends on the export target?

**Comment by reknih** [comment]
This uses C++ under the hood, we can only link to other pure Rust crates for WASM compilation. I'm too lazy to retrieve the issue right now but it has been thrown around here a few times.

**Comment by edrihan** [comment]
Just came here to say like others have that I started checking typst out because it seems really great and simpler than LaTeX.. but I have a problem that I'm want to share assets with LaTeX and I have 100's of GB of assets that I may use.. so having svg versions of everything isn't very practical. I will be keeping en eye on this project and hopefully one day it will be possible/financially viable to have pdf support and be a real contender for LaTeX replacement (for me and others in a similar situation)! As much as it may be inconvenient this seems like kind of an essential feature, as pdf is the dominant document format in the world. The solution written by @balping demonstrates that it's possible to do. That approach could be merged? It would make a superior binary file if it was like that. I think for most people.. you don't need the webapp.. because I can't have my other python files and assets on there anyway! So if you're gonna do custom workflows you need to use your own computer anyway. Why punish the computer (pdf) users (which I don't think are insignificant) just because pure-Rust pdf rendering isn't available yet. People could start replacing LaTeX in their workflows on their computers and maybe later if someone solves the issue for web then it could be adopted at that point? n=1, but this makes Typst unusable in my use-case as a document creation system. Really sad because I found out I can do outlined rainbow gradient text which is hard to do in latex... Damned if you use LaTeX, damned if you don't use Typst because it doesn't support the "protected **document format**"

**Comment by Kreijstal** [comment]
> Just came here to say like others have that I started checking typst out because it seems really great and simpler than LaTeX.. But - I have a problem: that I'm want to share assets with LaTeX and I have 100's of GB of assets that I may use.. so having to make svg versions of everything isn't very practical. I will be keeping en eye on this project and hopefully one day it will be possible/financially viable to have pdf support, and be a real contender for LaTeX replacement (for me and others in a similar situation)!
> 
> As much as it may be inconvenient, this seems like kind of an **essential** feature, as pdf is the dominant document format in the world. The solution written by @balping demonstrates that it's possible to do. That approach could be merged? It would make a superior binary file if it was like that. I think for most people.. you don't need the webapp.. because I can't have my other python files and assets on there anyway! So if you're gonna do custom workflows you need to use your own computer anyway. Why punish the computer (pdf) users (which I don't think are insignificant) just because pure-Rust pdf rendering isn't available yet. People could start replacing LaTeX in their workflows on their computers and maybe later if someone solves the issue for web then it could be adopted at that point?
> 
> n=1, but this makes Typst unusable in my use-case as a document creation system. Really sad because I found out I can do outlined rainbow gradient text which is hard to do in latex. And holy cow wouldn't it be nice to use a real coding language instead of latex arcane "wizard consultation services" like tex.stackexchange?... But it seems we are still "damned if you [use LaTeX], damned if you don't [use Typst]" because it doesn't support the most common format used by documents in the modern age.
> 
> I really hope that this isn't forgotten abourt.. because this could really put typst on the map for a lot of people.. and there are even a lot of people in this thread who stated how important this really is. My 2c as a disgruntled LaTeX user who wants to use something better.
> 
> P.s. I think typst is really great in theory!

I mean you could just the fork instead no?

**Comment by siefkenj** [comment]
I don't know when this started to be the case, but it appears mupdf now has a WASM implementation https://mupdf.readthedocs.io/en/latest/mupdf-wasm.html

@balping could your approach be modified now to be a regular typst plugin?

**Comment by Enivex** [comment]
> I don't know when this started to be the case, but it appears mupdf now has a WASM implementation https://mupdf.readthedocs.io/en/latest/mupdf-wasm.html
> 
> @balping could your approach be modified now to be a regular typst plugin?

That has been the case for a long time. It's not the kind of wasm that can be used for typst.

**Comment by francois-rozet** [comment]
> we are not willing to add an image format that won't work in all export targets

I understand that adding an exception for the PDF format is not "elegant", but the overwhelming majority of users are compiling to PDF, not PNG, SVG or HTML. As many have said before, without this feature, journals, conferences, and therefore authors, are unlikely to ever adopt Typst.

You are greatly limiting the reach of your own (fantastic) work by refusing to implement this feature, and for reasons that appear pointless to me (not creating an exception).

**Comment by Enivex** [comment]
Anyone using the web app or tinymist for preview are not using pdf as a target, for one.

**Comment by Dherse** [comment]
One way I could see us supporting it to a limited extent would be to have a special case in the WASM plugin system to allow running the WASM bundle produced by mupdf-wasm. It would still export to SVG and import SVGs into the PDF which is limiting but it could be done at least. The problem is that it would require supporting dozens of specific emscripten functions in the WASM runtime just for it. While not impossible, it is very dirty.

**Comment by francois-rozet** [comment]
If I understand https://github.com/typst/typst/issues/145#issuecomment-1565033965 correctly, a PDF can be embedded into another without intermediate export to SVG. This would not require MuPDF (potentially less work), but it would only work for PDF export (which I argue is fine). Therefore, there are 3 choices until a pure-rust PDF renderer exists:

1. No PDF image support.
2. PDF image support only for PDF export.
3. Hacky PDF image support using MuPDF-WASM.

**Comment by Dherse** [comment]
Yes, but that is unlikely to happen, the solution I am suggesting keep compatibility with all existing export targets, does not require any weird PDF embedded or worse PDF reading/handling and is similarly simple. A few hundreds lines in the Plugin section of Typst to get access to a limited subset of this feature.

I would argue that a solution that only works for PDF is not find as it breaks the web app (which is the core business of Typst the company with which I am not affiliated in any way). The solution I am offering gives limited compatibility (no text selection) but otherwise feature complete embedding.

**Comment by francois-rozet** [comment]
> I would argue that a solution that only works for PDF is not find as it breaks the web app

If it indeed breaks the web app, I understand the reluctance. However, I don't see how it would break the web app. There would be a feature missing from the web app, but that does not make it less usable.

**Comment by Moelf** [comment]
it's clearly not idea but what about making web app a subset in terms of functionality -- I think people who have large / immovable need for pdf embeddings are likely going to run Typst locally in a manual loop with whatever program that's generating hundreds of PDFs. You can frame PDF embedding as a, idk, professional or legacy compatible feature that's only accessible locally and not via the web app since your typical/modern/typst-native/web-native users won't need that so they just don't worry about it.

**Comment by Andrew15-5** [comment]
> If it indeed breaks the web app, I understand the reluctance. However, I don't see how it would break the web app. There would be a feature missing from the web app, but that does not make it less usable.

If it's "missing some features" then it literally makes it "less usable" as a whole. It of course doesn't make it less usable if a user doesn't use the missing features, but such users definitely can't represent the whole group of web app users.

**Comment by Andrew15-5** [comment]
> it's clearly not idea but what about making web app a subset in terms of functionality -- I think people who have large / immovable need for pdf embeddings are likely going to run Typst locally in a manual loop with whatever program that's generating hundreds of PDFs.

Although it's indeed _likely_ for documents that depends on generated PDF files for embedding to be compiled locally, but it's still only _likely_ and not _definite_. Furthermore, not all files must be generated (many times or at all), which makes it possible (and tempting for some folks) to use the web app for this use case. 

> You can frame PDF embedding as a, idk, professional or legacy compatible feature that's only accessible locally and not via the web app since your typical/modern/typst-native/web-native users won't need that so they just don't worry about it.

Typst is meant for different use cases, including "professional" documents. And some typical users are also former/current LaTeX users, so cutting them off with this CLI-only feature wouldn't be nice.

**Comment by Ghost4Man** [comment]
As a user of the web app who also wants embedded PDFs (with selectable and searchable text, etc.), I could imagine that when uploading a PDF into the web app, **in addition to converting it to SVG** (like it does today), it would (have an option to) **keep the original PDF around** specifically only for PDF export, but **use the SVG for everything else**. Or even somehow embed the SVG preview _inside_ the PDF. Would that be viable?

**Comment by Moelf** [comment]
>it literally makes it "less usable" as a whole

Currently there no support of embedding PDF in web or local, adding this feature for local usage doesn't make the web app less usable than it already is

**Comment by Enivex** [comment]
Please keep in mind that every time you reply to this topic everyone who hasn't actively unsubscribed will be notified. Rehashing the same arguments over and over isn't particularly helpful.

The team is aware of all of this, and it even came up in the recent community meeting (see the Discord).

**Comment by siefkenj** [comment]
I have been attempting to get the mupdf WASM build to run in `wasmi`, the WASM interpreter. If the appropriate stubs could be made, this whole bundle could be complied to WASM and run as a plugin in the current system.

**Comment by tobiasdiez** [comment]
@siefkenj Wasmi now has wasi support via [wasmi_wasi](https://github.com/wasmi-labs/wasmi/blob/main/crates/wasi). Perhaps its easier to add support for this in typst than adding more stubs.

**Comment by Kreijstal** [comment]
doesn't lopdf does something about that?

**Comment by siefkenj** [comment]
> @siefkenj Wasmi now has wasi support via [wasmi_wasi](https://github.com/wasmi-labs/wasmi/blob/main/crates/wasi). Perhaps its easier to add support for this in typst than adding more stubs.

I don't know if that is helpful. It appears mupdf is compiled via emscripten which targets the browser and not wasi. So we need a compatibility layer for all the emscripten functions. I've been searching and haven't found one, but I would love it if someone discovered one!

**Comment by Enivex** [comment]
This has also been discussed before. It's not currently possible to compile mupdf as a typst plugin.

**Comment by siefkenj** [comment]
> This has also been discussed before. It's not currently possible to compile mupdf as a typst plugin.

Yes, not directly. But `wasmi` can directly compile to wasm itself. So, it may be possible to have a wasm complied wasm interpreter, where the "higher level" wasm works as a plugin.

**Comment by Kreijstal** [comment]
> > This has also been discussed before. It's not currently possible to compile mupdf as a typst plugin.
> 
> Yes, not directly. But `wasmi` can directly compile to wasm itself. So, it may be possible to have a wasm complied wasm interpreter, where the "higher level" wasm works as a plugin.

that might be crazy enough to just work

**Comment by laurmaedje** [comment]
I can't comment on whether mupdf can somehow be coerced into running in a Typst plugin, but I'd like to note that in any case that could only be a Typst Universe package and not built-in because mupdf is AGPL licensed.

**Comment by frozolotl** [comment]
I just released MuchPDF, a package that can embed PDF images into Typst documents:
- https://github.com/frozolotl/muchpdf
- https://typst.app/universe/package/muchpdf

It works by stubbing a number of functions in emscripten's libc. This includes `setjmp` and `longjmp`, which means that errors are pretty bad.
It also can't handle large PDFs well (though there is a workaround given in the README). I'd definitely like to improve on that.

Even with these caveats, I think it can be a decent solution for most users.

**Comment by drskoolie** [comment]
## The One Compromise on the Rendering Rule?

I've been following this topic for the past year, and religiously coming back and checking this page. I'm sure I'm not the only one. 

**Summary of Issue**: Many LaTeX users heavily rely on directly embedding PDFs into figures. _This isn't supported in Typst._
**What's stopping PDF inserts???**: There isn't a library that exists currently that allows you to render PDF inserts to the following types: _pdf_, _webapp_, _png_, (and in the future _html_).
**Proposed Workaround from the Typst team**: Convert the PDF to SVG, and then embed the SVG instead. This works, but isn't optimal. Especially for larger sized PDFs.
**Current State of Libraries:** There exists libraries that allow you to render PDF inserts to _pdf_ only. 
**Compromise:** Allow embedding of PDFs, and have it render correctly if you are exporting PDF. But for all the other types (webapp, svg), a banner shows that says that this feature is only supported by rendering the document to PDF. 


This will be a temporary compromise until the libraries are written to handle conversion into the other formats. The only reason I'm proposing this is because it's a heavily used feature in LaTeX and I see it as the only thing holding back the mass adoption of Typst by LaTeX users.

**Comment by earmer** [comment]
In PDF.js, the author adds a text layer at the top and sets it to be invisible to make the text selectable. Typst can also do this: convert to SVG, extract text and location, and attach a text layer to it.

**Comment by Kreijstal** [comment]
> In PDF.js, the author adds a text layer at the top and sets it to be invisible to make the text selectable. Typst can also do this: convert to SVG, extract text and location, and attach a text layer to it.

How to do it in c/rust/wasm?

**Comment by earmer** [comment]
@Kreijstal 
Well, in python we can use pdfplumber. Like
```python
import pdfplumber

pdf = pdfplumber.open('t.pdf')
page = pdf.pages[4]

words = page.extract_words()
im = page.to_image()
im.draw_rects(words)
```

As for Rust/C. 
https://gist.github.com/earmer/f4b04370c3a2317105d5ab0ffea2ffcb

Might help.

**Comment by noahfraiture** [comment]
> ## The One Compromise on the Rendering Rule?
> I've been following this topic for the past year, and religiously coming back and checking this page. I'm sure I'm not the only one.
> 
> **Summary of Issue**: Many LaTeX users heavily rely on directly embedding PDFs into figures. _This isn't supported in Typst._ **What's stopping PDF inserts???**: There isn't a library that exists currently that allows you to render PDF inserts to the following types: _pdf_, _webapp_, _png_, (and in the future _html_). **Proposed Workaround from the Typst team**: Convert the PDF to SVG, and then embed the SVG instead. This works, but isn't optimal. Especially for larger sized PDFs. **Current State of Libraries:** There exists libraries that allow you to render PDF inserts to _pdf_ only. **Compromise:** Allow embedding of PDFs, and have it render correctly if you are exporting PDF. But for all the other types (webapp, svg), a banner shows that says that this feature is only supported by rendering the document to PDF.
> 
> This will be a temporary compromise until the libraries are written to handle conversion into the other formats. The only reason I'm proposing this is because it's a heavily used feature in LaTeX and I see it as the only thing holding back the mass adoption of Typst by LaTeX users.

Hello,
I've been writing a paper using Typst and recently discovered that significant issue that I believe hinders broader adoption. The proposed compromise seems promising, as it appears that users who want to embed PDFs without converting them to SVG are primarily those who need to export to PDF. 
Is this issue being taken seriously by the maintainers, or is it considered an idea unlikely to be implemented in Typst? 
We are two years after this issue has been open and 4 months after your message, and it seems that nothing changed.

**Comment by Kreijstal** [comment]
> > ## The One Compromise on the Rendering Rule?
> > I've been following this topic for the past year, and religiously coming back and checking this page. I'm sure I'm not the only one.
> > **Summary of Issue**: Many LaTeX users heavily rely on directly embedding PDFs into figures. _This isn't supported in Typst._ **What's stopping PDF inserts???**: There isn't a library that exists currently that allows you to render PDF inserts to the following types: _pdf_, _webapp_, _png_, (and in the future _html_). **Proposed Workaround from the Typst team**: Convert the PDF to SVG, and then embed the SVG instead. This works, but isn't optimal. Especially for larger sized PDFs. **Current State of Libraries:** There exists libraries that allow you to render PDF inserts to _pdf_ only. **Compromise:** Allow embedding of PDFs, and have it render correctly if you are exporting PDF. But for all the other types (webapp, svg), a banner shows that says that this feature is only supported by rendering the document to PDF.
> > This will be a temporary compromise until the libraries are written to handle conversion into the other formats. The only reason I'm proposing this is because it's a heavily used feature in LaTeX and I see it as the only thing holding back the mass adoption of Typst by LaTeX users.
> 
> Hello, I've been writing a paper using Typst and recently discovered that significant issue that I believe hinders broader adoption. The proposed compromise seems promising, as it appears that users who want to embed PDFs without converting them to SVG are primarily those who need to export to PDF. Is this issue being taken seriously by the maintainers, or is it considered an idea unlikely to be implemented in Typst? We are two years after this issue has been open and 4 months after your message, and it seems that nothing changed. Relying exclusively on workaround isn't the best

just make a script pdf2svg.sh, I wouldnt hold my breath

**Comment by Dherse** [comment]
It's not so much a workaround as the current intended way. 

Currently embedding PDF causes limitations to compatibility especially with tools like the web app preview (the thing that keeps typst the company alive and thriving which ends up being good for all typst users). Rendering PDFs to SVG of PNG for other targets would require a reader that is compatible with the build pipeline that typst uses (especially for WASM) and compatible with the license model that typst uses. The web app already supports conversions when importing PDF files on the fly.

And I would say that saying "I would not hold my breath for it" about an open source projects where countless contributors and the creators have spent thousand of hours cumulatively making a free and open source tool can sound a bit diminishing. Although it is true that it's unlike to happen in the coming months. It's not due to lack of effort or disinterest in having this feature supported properly.

**Comment by siefkenj** [comment]
Just a guess, but if someone pooled enough money to support Typst development for ~6months, I suspect that the core developers would elevate the priority of this bug https://typst.app/pricing/  

Save that, I would try to use @frozolotl plugin (I believe there are licensing issues with including it in the core of Typst..?).

**Comment by Andrew15-5** [comment]
I need to clarify that https://github.com/typst/typst/issues/145#issuecomment-2562005379 which is being quoted is misleading as a very promising and _probably_ "good enough" solution for the majority of people was written right above it: https://github.com/typst/typst/issues/145#issuecomment-2559968887.

Try https://typst.app/universe/package/muchpdf, if it doesn't work, check FAQ section at the bottom (of its README/description). If still have problems with using it, you can open an issue in https://github.com/frozolotl/muchpdf/issues.

**Comment by Dherse** [comment]
Sorry to say that even in six months you would not have a complete PDF reading library. More like pooled the resources to get typst GMBH the financial ability to get a license... why? PDF is a hell of a standard, I once discussed with some of the mupdf devs on their discord and they gave me a few examples that show there are many many PDF writers out there that make a bit of a mess and good readers (like mupdf) can kind of recover from those and still display a document.

**Comment by earmer** [comment]
I just wonder, is it really necessary for anyone to use PDF figures? I
really want to know the names of these software programs that cannot output
SVGs but can output PDFs.

**Comment by tractatus** [comment]
@earmer not softwares but the academic publishing industry. 
Figures are supplied as PDF in all faces of submission, never as SVG.

**Comment by laurmaedje** [comment]
This issue has a lot of people subscribed to it, so I would ask everyone to not leave comments that bring nothing new to the table. I would prefer not to lock it, but pretty much everything has already been said in the thread above.

We're aware that many people consider this issue important. That said, our fundamental stance on this hasn't changed --- we're happy to support it once an apprioriate solution for HTML/PNG/SVG export is available (most likely a minimal PDF rendering library maintained by us). But this is a lot of work and there is no timeframe for this at the moment. In the meantime, conversion to SVG is a fairly good workaround, in the form of automatic conversion in the web app, through the muchpdf package, or manually.

**Comment by jheuel** [comment]
I use a small function to save the dimensions of all pdf images and then replace them in the rendered output using pymupdf.

```typst
[#muchpdf(read(fn, encoding: none), pages: page) <__pdfembed>]

context [
  #let embed = query(selector(<__pdfembed>).before(here())).last()
  #layout(size => [
    #let dim = measure(block(width: size.width, embed))
    #let pos = embed.location().position()
    #metadata((
      filename: fn,
      src_page: page,
      page: pos.page - 1,
      x: pos.x.pt(),
      y: pos.y.pt(),
      width: dim.width.pt(),
      height: dim.height.pt(),
      overlay: overlay,
    )) <__pdfembedmetadata>
  ])
]

```
Full code is here: https://github.com/jheuel/typst_embed_pdf

**Comment by balping** [comment]
@jheuel 

This is very similar to [my code](https://gitlab.com/balping/typst-pdf-embedder) , see above: https://github.com/typst/typst/issues/145#issuecomment-2336794215

**Comment by jheuel** [comment]
@balping 

Oh, did not see that message otherwise I would not have looked into it.

I also started with drawn rectangles and then redacting them in PyMuPDF but could not get it to work because I want to use pdf plots in slides with some annotations on top and the redacting just nukes everything in the rectangle. You can also skip the redacting step if you draw hidden rectangles.

**Comment by LaurenzV** [comment]
Hey guys! I've been working on a PDF renderer in the last few weeks/months, and while there still are a few bugs I'm aware of as well as a few features that definitely need to be implemented, I'm hoping that it supports enough features that it should already be able to render a very considerable chunk of PDF documents. It would be very helpful to have it tested with more real-life PDFs (especially the kind you would embed in a Typst document).

So, if you want to help and have some time on your hand as well as PDFs, you can test them using this very small demo tool I built: https://laurenzv.github.io/hayro/ (this is very bare-bones and intentionally doesn't support features like zooming, the main point is to be able to view the pages of a PDF to identify PDFs that either cause a crash  or render wrongly), and if you find something broken you can open an issue at https://github.com/LaurenzV/hayro/issues. This would help me identify what issues are more common and need prioritization.

Thank you!

**Comment by sijow** [comment]
It works great with the TikZ, Mathematica, Matlab and matplotlib figures I have tried.

**Comment by Andrew15-5** [comment]
Does this rasterizer actually only supports RGBA8 encoded raster images, or does it supposed to also render to SVG like MuchPDF?

**Comment by LaurenzV** [comment]
Right now only RGBA8, but it is possible to add SVG later on.

**Comment by nico** [comment]
> So, if you want to help and have some time on your hand as well as PDFs

Hello! I hack on SerenityOS's LibPDF, and someone on our discord pointed me at your comment. I thought I'd do a short writeup on how we do PDF compat testing and bringup, in the hope that it's useful, if you want to be a bit more proactive about finding PDFs that don't work.

SerenityOS doesn't use any external dependencies, so all support code called by LibPDF (font file parsing, image decoders, etc) is also in SerenityOS's repository. It tends to be written for understandability and spec compliance more than for raw performance, so maybe the code is a useful resource too. But this is about testing, so onwards:

Our PDF rendering code returns error codes on things that aren't implemented yet (example: https://github.com/SerenityOS/serenity/blob/6891956da43dc6d53d8b8cc7c73ec50de654b2e1/Userland/Libraries/LibPDF/Renderer.cpp#L741)

When rendering a page, we collect all the errors we see and return them (https://github.com/SerenityOS/serenity/blob/6891956da43dc6d53d8b8cc7c73ec50de654b2e1/Userland/Libraries/LibPDF/Renderer.cpp#L111).

We have a `pdf` binary that can render a given page of a PDF, and it prints all the errors it sees rendering that page to stdout. That looks like so:

```
% Build/lagom/bin/pdf --render out.png ~/Downloads/pdf_reference_1-7.pdf --page 1143
...
warning: Rendering of feature not supported: Pattern color spaces not yet implemented
warning: Rendering of feature not supported: Pattern color spaces not yet implemented
warning: Rendering of feature not supported: Pattern color spaces not yet implemented
warning: Rendering of feature not supported: Pattern color spaces not yet implemented
warning: Rendering of feature not supported: Pattern color spaces not yet implemented
warning: Rendering of feature not supported: Pattern color spaces not yet implemented
warning: Rendering of feature not supported: Pattern color spaces not yet implemented
warning: Rendering of feature not supported: Pattern color spaces not yet implemented
warning: Rendering of feature not supported: Pattern color spaces not yet implemented
warning: Rendering of feature not supported: Pattern color spaces not yet implemented
```

`pdf` also has a mode where it renders all pages of a PDF to memory, collects all these diagnostics, aggregates them, and prints the aggregate (https://github.com/SerenityOS/serenity/blob/6891956da43dc6d53d8b8cc7c73ec50de654b2e1/Userland/Utilities/pdf.cpp#L131):

```
% Build/lagom/bin/pdf --debugging-stats ~/Downloads/pdf_reference_1-7.pdf
...
84 times: Rendering of feature not supported: Pattern color spaces not yet implemented
    on pages: 831 (2x) 832 (2x) 1143 (10x) 1144 (8x) 1145 (8x) 1146 (4x) 1149 (34x) 1150 (16x)
```

If you also pass `--json`, it prints the same in json format:

```
{"num_pages":1310,"issues":{"Rendering of feature not supported: Pattern color spaces not yet implemented":[[831,2],[832,2],[1143,10],[1144,8],[1145,8],[1146,4],[1149,34],[1150,16]]}}
```

(That particular file has only one diagnostic left.)

We then have a python script that takes the name of a directory and runs `pdf --debugging-stats --json` on all pdf files in that directory and aggregates those and prints a report: https://github.com/SerenityOS/serenity/blob/6891956da43dc6d53d8b8cc7c73ec50de654b2e1/Meta/test_pdf.py

https://digitalcorpora.org/corpora/file-corpora/cc-main-2021-31-pdf-untruncated/ has 8 million PDFs downloaded from the internet, available in zips of 1000 files sorted by sha1 (i.e. effectively random subsets – but they're from a crawl from 2021, so they skew towards being new). We then run `time Meta/test_pdf.py ~/Downloads/0000`, where that folder is the first 1000 files from that test corpus. Here's a recent example output:

[pdf-report-84.txt](https://github.com/user-attachments/files/20969076/pdf-report-84.txt)

I have an archive of those over time, which lets me track how things are improving by diffing consecutive runs. This also makes sure nothing regresses too badly.

(We currently don't have rendering tests; eventually we'll want to add those too.)

https://github.com/pdf-association/pdf-corpora?tab=readme-ov-file links to more PDF test corpora. Other PDF engines also have a bunch of test files can be tested with.

SerenityOS has a few PDF test files, for the most part hand-written by me: https://github.com/SerenityOS/serenity/tree/6891956da43dc6d53d8b8cc7c73ec50de654b2e1/Tests/LibPDF

I think this approach has worked fairly well – but to be fair, we still don't support a few features (but the end for support of the basic feature set is in sight). Many things do work, and in a few cases we are more spec-compliant than all other PDF viewers I've tried (example: https://github.com/SerenityOS/serenity/pull/25955#issuecomment-2906005042)

Again, I hope this is useful. Feel free to ignore if it isn't :)

**Comment by LaurenzV** [comment]
It is useful, thanks a lot! I've actually already stumbled upon your handwritten tests before, and some of them are in my test suite. :) I might try going through one of the bigger corpora in the future, but for now I'm focusing on the pdf.js/PDFBox test suite, as well as any PDFs users here might have, I think this will already keep me very busy for a while. :D

**Comment by davystrong** [comment]
I tested a [PDF](https://github.com/user-attachments/files/20981274/metric_view.pdf) with an embedded PNG and it just outputted noise. The PNG is RGB8, rather than ARGB8. I assume this is the reason? Or should I create an issue?

**Comment by LaurenzV** [comment]
Yep, please create an issue and I'll take a look when I find the time, thanks!

**Comment by laurmaedje** [comment]
Thanks to the valiant work by @LaurenzV, the development build of Typst now supports embedding PDFs as images (in all export formats)!

To make sure the feature works as robustly as possible once 0.14 comes out, we'd like to ask for help testing it.

You can test it locally or in the web app. To test locally, install the latest commit via `cargo`. To test in the web app, simply select "Development version (July 22, 2025)" under "Compiler version" in the settings side panel.

**Comment by andreas-bulling** [comment]
@laurmaedje What is the command required to do so?

**Comment by sijow** [comment]
Just `#image("file.pdf")`
